store/model.go

```diff
@@ -0,0 +1,219 @@
+	upsert operationType = "upsert"
+	delete operationType = "delete"
+)
+
+type Model struct {
+	lock            sync.RWMutex
+	schema          *jsonschema.Schema
+	valueType       reflect.Type
+	datastore       ds.Datastore
+	dispatcher      *eventstore.Dispatcher
+	dispatcherToken eventstore.Token
+}
+
+func (m *Model) Read(f func(txn *Txn) error) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	txn := &Txn{model: m, readonly: true, ops: make(map[ds.Key]operation)}
+	defer txn.Discard()
+	if err := f(txn); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *Model) Update(f func(txn *Txn) error) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	txn := &Txn{model: m, ops: make(map[ds.Key]operation)}
+	defer txn.Discard()
+	if err := f(txn); err != nil {
+		return err
+	}
+	return txn.Commit()
+}
+
+func (m *Model) FindByID(id string, v interface{}) error {
+	key := ds.NewKey(id)
+	bytes, err := m.datastore.Get(key)
+	if errors.Is(err, ds.ErrNotFound) {
+		return ErrNotFound
+	}
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(bytes, v)
+}
+
+func (m *Model) Reduce(event eventstore.Event) (err error) {
+	log.Debugf("reducer %s start", m.schema.Ref)
+	defer log.Debugf("reducer %s end with err: %v", m.schema.Ref, err)
+	if event.Type() != m.schema.Ref {
+		log.Debugf("ignoring event from uninteresting type")
+		return nil
+	}
+	var op operation
+	err = json.Unmarshal(event.Body(), &op)
+	if err != nil {
+		return
+	}
+
+	key := ds.NewKey(event.EntityID())
+	switch op.Type {
+	case upsert:
+		var value, patchedValue []byte
+		value, err = m.datastore.Get(key)
+		if errors.Is(err, ds.ErrNotFound) {
+			if err = m.datastore.Put(key, op.JSONPatch); err != nil {
+				return
+			}
+			log.Debug("\tinsert operation applied")
+			return
+		}
+		if err != nil {
+			return
+		}
+		patchedValue, err = jsonpatch.MergePatch(value, op.JSONPatch)
+		if err != nil {
+			return fmt.Errorf("error when patching value: %v", err)
+		}
+		if err = m.datastore.Put(key, patchedValue); err != nil {
+			return
+		}
+		log.Debug("\tupdate operation applied")
+	case delete:
+		if err = m.datastore.Delete(key); err != nil {
+			return
+		}
+		log.Debug("\tdelete operation applied")
+	default:
+		return fmt.Errorf("unknown operation %s", op.Type)
+	}
+
+	return nil
+}
+
+type Txn struct {
+	model     *Model
+	discarded bool
+	commited  bool
+	readonly  bool
+	ops       map[ds.Key]operation
+}
+
+type operation struct {
+	Type      operationType
+	EntityID  string
+	JSONPatch []byte
+}
+
+func (t *Txn) Discard() {
+	t.discarded = true
+}
+
+func (t *Txn) Commit() error {
+	if t.discarded || t.commited {
+		return fmt.Errorf("can't commit discarded/commited txn")
+	}
+	now := time.Now()
+	//  ToDo/Important: As first approximation, each key change is a separate event
+	for _, op := range t.ops {
+		opBytes, err := json.Marshal(op)
+		if err != nil {
+			return err
+		}
+		event := eventstore.NewJsonPatchEvent(now, op.EntityID, t.model.schema.Ref, opBytes)
```

Comment:
I think here we want to add this same information as the body of a thread event, which will in turn dispatch a full record that can be picked up by reducers. That is to say, the dispatching can happen internal to threads.

Reply:
(just flagging for my own comprehension here)

Reply:
I guess you're refering to the discussion in the Slack-thread (maybe your comment happened before :)); I mean the "should eventstore Events leak extra info to thread Events structures" thing.
I kind of like moving towards a thread-Event structure which is very opaque (such as simply `[]byte`) since adding extra metadata at that level may have a purpose for this application, but other users that will use directly `go-textile-threads` without eventstore, won't. 
Maybe is what you're saying (and I'm kind of repeating myself as in Slack :stuck_out_tongue: )

Reply:
Yep, agreed